code/espurna/ota_httpupdate.ino

```diff
@@ -0,0 +1,244 @@
+        case HTTP_UPDATE_FAILED:
+            DEBUG_MSG_P(PSTR("[OTA] Update failed (error %d): %s\n"), ESPhttpUpdate.getLastError(), ESPhttpUpdate.getLastErrorString().c_str());
+            eepromRotate(true);
+            break;
+        case HTTP_UPDATE_NO_UPDATES:
+            DEBUG_MSG_P(PSTR("[OTA] No updates"));
+            eepromRotate(true);
+            break;
+        case HTTP_UPDATE_OK:
+            DEBUG_MSG_P(PSTR("[OTA] Done, restarting..."));
+            deferredReset(500, CUSTOM_RESET_OTA); // wait a bit more than usual
+            break;
+    }
+
+}
+
+#if OTA_CLIENT_HTTPUPDATE_2_3_0_COMPATIBLE
+void _otaClientFromHttp(const String& url) {
+    _otaClientRunUpdater(nullptr, url, "");
+}
+#else
+void _otaClientFromHttp(const String& url) {
+    auto client = std::make_unique<WiFiClient>();
+    _otaClientRunUpdater(client.get(), url, "");
+}
+#endif
+
+#if SECURE_CLIENT == SECURE_CLIENT_BEARSSL
+
+void _otaClientFromHttps(const String& url) {
+
+    int check = getSetting("otaScCheck", SECURE_CLIENT_CHECK_FINGERPRINT).toInt();
+    bool settime = (check == SECURE_CLIENT_CHECK_CA);
+
+    if (!ntpSynced() && settime) {
+        DEBUG_MSG_P(PSTR("[OTA] Time not synced!\n"));
+        return;
+    }
+
+    // unique_ptr self-destructs after exiting function scope
+    // create WiFiClient on heap to use less stack space
+    auto client = std::make_unique<BearSSL::WiFiClientSecure>();
+
+    if (check == SECURE_CLIENT_CHECK_NONE) {
+        DEBUG_MSG_P(PSTR("[OTA] !!! Connection will not be validated !!!\n"));
+        client->setInsecure();
+    }
+
+    if (check == SECURE_CLIENT_CHECK_FINGERPRINT) {
+        String fp_string = getSetting("otafp", OTA_FINGERPRINT);
+        if (!fp_string.length()) {
+            DEBUG_MSG_P(PSTR("[OTA] Requested fingerprint auth, but 'otafp' is not set\n"));
+            return;
+        }
+
+        uint8_t fp_bytes[20] = {0};
+        sslFingerPrintArray(fp_string.c_str(), fp_bytes);
+
+        client->setFingerprint(fp_bytes);
+    }
+
+    BearSSL::X509List *ca = nullptr;
+    if (check == SECURE_CLIENT_CHECK_CA) {
+        ca = new BearSSL::X509List(_ota_client_http_update_ca);
+        // because we do not support libc methods of getting time, force client to use ntpclientlib's current time
+        // XXX: local2utc method use is detrimental when DST happening. now() should be utc
+        client->setX509Time(ntpLocal2UTC(now()));
+        client->setTrustAnchors(ca);
+    }
+
+    // TODO: provide terminal command for probing?
+    // TODO: RX and TX buffer sizes must be equal?
+    uint16_t requested_mfln = getSetting("otaScMFLN", OTA_SECURE_CLIENT_MFLN).toInt();
+    if (requested_mfln) {
+        URL _url(url);
+        bool supported = client->probeMaxFragmentLength(_url.host.c_str(), _url.port, requested_mfln);
+        DEBUG_MSG_P(PSTR("[OTA] MFLN buffer size %u supported: %s\n"), requested_mfln, supported ? "YES" : "NO");
+        if (!supported) return;
```

Comment:
Instead of returning, shouldn't this code be:
```c++
if (supported) {
  client->setBufferSizes(requested_mfln, requested_mfln);
}
```
Because now it never calls ``_otaClientRunUpdater()`` if MFLN is not supported.

Reply:
But why ignore `requested_mfln`? Default is off, so it either works or not.

>  RX and TX buffer sizes must be equal?

Looking at this now, does probing check how much server can send or receive? Do buffers need symmetry?

Reply:
The question is if the function should fail if MFLN is not supported by the server. I think it's more of a "nice to have" thing, and the function should just proceed if MFLN is not supported?

Reply:
Maybe I am overthinking it. It would be nice to not blow up memory though?
Or client can have a "nice" mode, probing all of the possible buffer sizes and settling on some and saving it as a setting.

Reply:
Actually probing by default might be a good idea - but each probe is of course some extra network traffic. For long-lived connections like MQTT I think it's definitely worth it to do by default.

The question is if TLS implementations (server-side) have limitations on the probe length. I'd think it MFLN is supported it will always support 512 and up, and not e.g. 4k only? So maybe 1 probe is enough? But hard to say without diving in OpenSSL/gnuTLS/OTP/etc code.